app/slave/cluster_slave.py

```diff
@@ -74,24 +75,36 @@ def _start_heartbeat(self):
 
     def _run_heartbeat(self):
         try:
-            self._send_heartbeat_to_master()
-            self._heartbeat_failure_count = 0
+            response = self._send_heartbeat_to_master()
+
+            # A non HTTP 200 response indicates that the master is up, but does not recognize the slave.
+            # The slave tries to reconnect to master at this point.
+            # This is also counted as a heartbeat failure and the slave takes an action accordingly
+            if response.status_code != http.client.OK:
+                self._act_on_heartbeat_failure()
```

Comment:
This method name is vague. If this method is doing so many things that "act" is the best descriptor, that's probably an indication that it's doing too many things.

Reply:
yeah.. It's a bad name. I'll think about a better name or possibly splitting it into multiple methods.

Reply:
I don't mean to imply that you can _always_ find a better name btw. Sometimes you you just have to make one method do 5 things. I don't think that's the case here though